### PR TITLE
subnet v2 resource (regionalized API)

### PR DESCRIPTION
**Pull request by AdrienPensart**
# Description

Add a new regionalized resource ovh_cloud_project_network_private_subnet_v2, it is not really a v2, but a regionalized subnet, so it may be preferable to call it ovh_cloud_project_network_private_subnet_on_region or something like this...

Related to https://github.com/ovh/public-cloud-roadmap/issues/582

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Improvement (improve existing resource(s) or datasource(s))
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A: ` make testacc TESTARGS="-run TestAccCloudProjectNetworkPrivateSubnetV2_basic"`

**Test Configuration**:
* Terraform version: `terraform version`: Terraform vx.y.z
* Existing HCL configuration you used: 
```hcl
resource "ovh_cloud_project_network_private_subnet_v2" "subnet" {
    service_name      = ovh_cloud_project_network_private.network.service_name
    name              = "my_new_subnet"
    network_id        = element(tolist(ovh_cloud_project_network_private.network.regions_attributes), 0).openstackid
    region            = element(tolist(ovh_cloud_project_network_private.network.regions), 0)
    cidr              = "192.168.0.0/24"
    dns_nameservers   = ["1.1.1.1"]
    host_route {
            destination = "192.168.0.0/24"
            nexthop = "192.168.0.254"
    }
    allocation_pool {
            start = "192.168.0.100"
            end = "192.168.0.200"
    }
    dhcp              = true
    gateway_ip        = "192.168.0.253"
    enable_gateway_ip = true
}
```

# Checklist:

- [ ] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings or issues
- [ ] I have added acceptance tests that prove my fix is effective or that my feature works
- [ ] New and existing acceptance tests pass locally with my changes
- [ ] I ran successfully `go mod vendor` if I added or modify `go.mod` file

### CODE REVIEW

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+	"github.com/ovh/terraform-provider-ovh/ovh/helpers"
+)
+
+func resourceOvhCloudProjectNetworkPrivateSubnetV2ImportState(
+	d *schema.ResourceData,
+	meta interface{}) ([]*schema.ResourceData, error) {
+	givenId := d.Id()
+	splitId := strings.SplitN(givenId, "/", 3)
+	if len(splitId) != 3 {
+		return nil, fmt.Errorf("Import Id is not service_name/network_id/subnet_id formatted")
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go:19]: region is missing from the required params right ?

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+func resourceCloudProjectNetworkPrivateSubnetV2() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudProjectNetworkPrivateSubnetV2Create,
+		Read:   resourceCloudProjectNetworkPrivateSubnetV2Read,
+		Delete: resourceCloudProjectNetworkPrivateSubnetV2Delete,
+		Importer: &schema.ResourceImporter{
+			State: resourceOvhCloudProjectNetworkPrivateSubnetV2ImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go:44]: missing description of many fields

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+			},
+			"gateway_ip": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+			},
+			"enable_gateway_ip": {
+				Type:     schema.TypeBool,
+				Optional: true,
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go:87]: required on the API

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: resourceCloudProjectNetworkPrivateSubnetValidateNetwork,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"gateway_ip": {
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go:79]: could use `resourceCloudProjectNetworkPrivateSubnetValidateIP` ?

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+					ValidateFunc: resourceCloudProjectNetworkPrivateSubnetValidateIP,
+				},
+				Description: "List of DNS nameservers, default: 213.186.33.99",
+			},
+			"host_routes": {
+				Type:     schema.TypeList,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeMap,
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go:106]: why don't you define the nested object ?

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+					Set:  schema.HashString,
+				},
+			},
+			"allocation_pools": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeMap,
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go:116]: same, you should define the real object

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+	config := meta.(*Config)
+	serviceName := d.Get("service_name").(string)
+	networkId := d.Get("network_id").(string)
+	regionName := d.Get("region").(string)
+
+	subnets := []*CloudProjectNetworkPrivateV2Response{}
+
+	log.Printf("[DEBUG] Will read public cloud private network subnet for project: %s, region: %s, network: %s, id: %s", serviceName, regionName, networkId, d.Id())
+
+	endpoint := fmt.Sprintf("/cloud/project/%s/region/%s/network/%s/subnet", serviceName, regionName, networkId)
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go:211]: why don't you just fetch the right subnet using `GET /cloud/project/{serviceName}/region/{regionName}/network/{networkId}/subnet/{subnetId}` ?

File: ovh/resource_cloud_project_network_private_subnet_v2_test.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+resource "ovh_cloud_project_network_private" "network" {
+  service_name = data.ovh_cloud_project_regions.regions.service_name
+  vlan_id    = 1
+  name       = "terraform_testacc_private_net"
+  regions    = slice(sort(tolist(data.ovh_cloud_project_regions.regions.names)), 0, 3)
+}
+`
+
+func testAccCloudProjectNetworkPrivateV2SubnetConfig(config string) string {
+	attachVrack := fmt.Sprintf(
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2_test.go:47]: `attachVrack` and `noAttachVrack` should each be declared only when necessary

File: ovh/resource_cloud_project_network_private_subnet_v2_test.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+					resource.TestCheckResourceAttr("ovh_cloud_project_network_private_subnet_v2.subnet", "dhcp", "true"),
+					resource.TestCheckResourceAttr("ovh_cloud_project_network_private_subnet_v2.subnet", "enable_gateway_ip", "true"),
+					resource.TestCheckResourceAttr("ovh_cloud_project_network_private_subnet_v2.subnet", "cidr", "192.168.169.0/24"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckcCloudProjectNetworkPrivateSubnetV2PreCheck(t *testing.T) {
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2_test.go:126]: func not needed, you should call each precheck on l102

File: website/docs/r/cloud_project_network_private_subnet_v2.html.markdown [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+* `region` - See Argument Reference above.
+* `gateway_ip` - See Argument Reference above.
+* `enable_gateway_ip` - See Argument Reference above.
+* `dns_nameservers` - See Argument Reference above.
+* `host_routes` - See Argument Reference above.
+* `allocation_pools` - See Argument Reference above.
+
+## Import
+
+Subnet in a private network of a public cloud project can be imported using the `service_name` , the `network_id` and the `subnet_id`, separated by "/" E.g.,
```
Comment (amstuta) [inline website/docs/r/cloud_project_network_private_subnet_v2.html.markdown:75]: missing region in params

**Review by amstuta — CHANGES_REQUESTED** [review]

File: ovh/resource_cloud_project_network_private_subnet_v2.go (hunk 5 above)
Comment (AdrienPensart) [reply ovh/resource_cloud_project_network_private_subnet_v2.go:106]: how ? it is not a standalone object in API / terraform, I already try with a full resource definition it does not work...

File: ovh/resource_cloud_project_network_private_subnet_v2.go (hunk 6 above)
Comment (AdrienPensart) [reply ovh/resource_cloud_project_network_private_subnet_v2.go:116]: how ? it is not a standalone object in API / terraform, I already try with a full resource definition it does not work...

File: ovh/resource_cloud_project_network_private_subnet_v2.go (hunk 3 above)
Comment (amstuta) [reply ovh/resource_cloud_project_network_private_subnet_v2.go]: to set as required

File: ovh/resource_cloud_project_network_private_subnet_v2_test.go [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+)
+
+func testAccCloudProjectNetworkPrivateV2SubnetConfig(config string) string {
+	var testAccCloudProjectNetworkPrivateSubnetV2Config_attachVrack = `
+
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2_test.go:13]:
```suggestion
```

File: ovh/resource_cloud_project_network_private_subnet_v2_test.go [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+	}
+`
+	attachVrack := fmt.Sprintf(
+		testAccCloudProjectNetworkPrivateSubnetV2Config_attachVrack,
+		os.Getenv("OVH_VRACK_SERVICE_TEST"),
+		os.Getenv("OVH_CLOUD_PROJECT_SERVICE_TEST"),
+	)
+
+	var testAccCloudProjectNetworkPrivateSubnetV2Config_noAttachVrack = `
+
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2_test.go:38]:
```suggestion
```

File: website/docs/r/cloud_project_network_private_subnet_v2.html.markdown [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+* `region` - See Argument Reference above.
+* `gateway_ip` - See Argument Reference above.
+* `enable_gateway_ip` - See Argument Reference above.
+* `dns_nameservers` - See Argument Reference above.
+* `host_routes` - See Argument Reference above.
+* `allocation_pools` - See Argument Reference above.
+
+## Import
+
+Subnet in a private network of a public cloud project can be imported using the `service_name` , `network_id`, `region` and the `subnet_id`, separated by "/" E.g.,
```
Comment (amstuta) [inline website/docs/r/cloud_project_network_private_subnet_v2.html.markdown:75]:
```suggestion
Subnet in a private network of a public cloud project can be imported using the `service_name` , `region`, `network_id` and the `subnet_id`, separated by "/" E.g.,
```

File: website/docs/r/cloud_project_network_private_subnet_v2.html.markdown [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+```bash
+$ terraform import ovh_cloud_project_network_private_subnet_v2.mysubnet ookie9mee8Shaeghaeleeju7Xeghohv6e/25807101-8aaa-4ea5-b507-61f0d661b101/0f0b73a4-403b-45e4-86d0-b438f1291909
```
Comment (amstuta) [inline website/docs/r/cloud_project_network_private_subnet_v2.html.markdown:78]: region is missing in the ID example

**Review by amstuta — CHANGES_REQUESTED** [review]

File: ovh/resource_cloud_project_network_private_subnet_v2.go [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Computed:     true,
+				Description:  "Gateway IP of subnet",
+				ValidateFunc: resourceCloudProjectNetworkPrivateSubnetValidateIP,
+			},
+			"enable_gateway_ip": {
+				Type:        schema.TypeBool,
+				Optional:    true,
```
Comment (amstuta) [inline ovh/resource_cloud_project_network_private_subnet_v2.go]:
```suggestion
				Required:    true,
```

**Review by amstuta — APPROVED** [review]

**Pull request merged**